velox/connectors/hive/storage_adapters/gcs/GCSFileSystem.cpp

```diff
@@ -0,0 +1,353 @@
+ public:
+  Impl(const Config* config) : gcsConfig_(config) {}
+
+  ~Impl() {}
+
+  // Use the input Config parameters and initialize the GCSClient.
+  void initializeClient() {
+    auto options = gc::Options{};
+    std::string scheme = gcsConfig_.scheme();
+    // if (scheme.empty()) scheme = "https";
+    if (scheme == "https") {
+      options.set<gc::UnifiedCredentialsOption>(
+          gc::MakeGoogleDefaultCredentials());
+    } else {
+      options.set<gc::UnifiedCredentialsOption>(gc::MakeInsecureCredentials());
+    }
+    options.set<gcs::UploadBufferSizeOption>(kUploadBufferSize);
+    if (!gcsConfig_.endpointOverride().empty()) {
+      options.set<gcs::RestEndpointOption>(
+          scheme + "://" + gcsConfig_.endpointOverride());
+    }
+    if (gcsConfig_.credentials()) {
+      options.set<gc::UnifiedCredentialsOption>(gcsConfig_.credentials());
+    }
+    client_ = std::make_shared<gcs::Client>(options);
+  }
+
+  gcs::Client* gcsClient() const {
+    return client_.get();
+  }
+
+ private:
+  const GCSConfig gcsConfig_;
+  std::shared_ptr<gcs::Client> client_;
+};
+
+GCSFileSystem::GCSFileSystem(std::shared_ptr<const Config> config)
+    : FileSystem(config) {
+  impl_ = std::make_shared<Impl>(config.get());
+}
+
+void GCSFileSystem::initializeClient() {
+  impl_->initializeClient();
+}
+
+std::unique_ptr<ReadFile> GCSFileSystem::openFileForRead(
+    std::string_view path,
+    const FileOptions& /*unused*/) {
+  const std::string gcspath = gcsPath(path);
+  auto gcsfile = std::make_unique<GCSReadFile>(gcspath, impl_->gcsClient());
+  gcsfile->initialize();
+  return gcsfile;
+}
+
+std::unique_ptr<WriteFile> GCSFileSystem::openFileForWrite(
+    std::string_view path,
+    const FileOptions& /*unused*/) {
+  VELOX_NYI();
+}
+
+void GCSFileSystem::remove(std::string_view path) {
+  if (!isGCSFile(path))
+    VELOX_FAIL("File {} is not a valid gcs file", path);
+  const std::string file = gcsPath(path);
+
+  // assumption it's a proper path
+  std::string bucket;
+  std::string object;
+  bucketAndKeyFromGCSPath(file, bucket, object);
+
+  if (!object.empty()) {
+    auto stat = impl_->gcsClient()->GetObjectMetadata(bucket, object);
+    VELOX_CHECK_GCS_OUTCOME(
+        stat.status(), "Failed to get metadata for GCS object", bucket, object);
+  }
+  auto ret = impl_->gcsClient()->DeleteObject(bucket, object);
+  VELOX_CHECK_GCS_OUTCOME(
+      ret, "Failed to get metadata for GCS object", bucket, object);
+}
+
+bool GCSFileSystem::exists(std::string_view path) {
+  std::vector<std::string> result;
+  if (!isGCSFile(path))
+    VELOX_FAIL("File {} is not a valid gcs file", path);
+  const std::string file = gcsPath(path);
+
+  // assumption it's a proper path
+  std::string bucket;
+  std::string object;
+  bucketAndKeyFromGCSPath(file, bucket, object);
+  using ::google::cloud::StatusOr;
+  StatusOr<gcs::BucketMetadata> metadata =
+      impl_->gcsClient()->GetBucketMetadata(bucket);
+
+  if (!metadata) {
```

Comment:
nit: return metadata != null ?

Reply:
metadata is an instance of StatusOr, but I will rewrite the condition.